src/libcrun/criu.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * Copyright (C) 2020 Adrian Reber <areber@redhat.com>
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE
+
+#include <config.h>
+
+#ifdef HAVE_CRIU
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <criu/criu.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include "container.h"
+#include "error.h"
+#include "status.h"
+#include "utils.h"
+
+#define CRIU_LOG_FILE "dump.log"
+
+int
+libcrun_container_checkpoint_linux_criu (libcrun_container_status_t *status,
+					 libcrun_container_t *container,
+					 libcrun_checkpoint_restore_t *
+					 cr_options, libcrun_error_t *err)
+{
+  runtime_spec_schema_config_schema *def = container->container_def;
+  cleanup_free char *path = NULL;
+  cleanup_close int image_fd;
+  cleanup_close int work_fd;
+  size_t i;
+  int ret;
+
+  if (geteuid ())
+    return crun_make_error (err, 0, "Checkpointing requires root");
+
+  /* No CRIU version or feature checking yet. In configure.ac there
+   * is a minimum CRIU version listed and so far it is good enough.
+   *
+   * The CRIU library also does not yet have an interface to CRIU
+   * the version of the binary. Right now it is only possible to
+   * query the version of the library via defines during buildtime.
+   *
+   * The whole CRIU library setup works this way, that the library
+   * is only a wrapper around RPC calls to the actual library. So
+   * if CRIU is updated and the SO of the library does not change,
+   * and crun is not rebuilt against the newer version, the version
+   * is still returning the values during buildtime and not from
+   * the actual running CRIU binary. The RPC interface between the
+   * library will not break, so no reason to worry, but it is not
+   * possible to detect (via the library) which CRIU version is
+   * actually being used. This needs to be added to CRIU upstream. */
```

Comment:
some time ago @0x7f454c46 [proposed](https://github.com/checkpoint-restore/criu/issues/744#issuecomment-513028275) the idea of adding the ability to check if a specific feature is available at runtime. This may be more useful/accurate information than relying only on the CRIU version.

Reply:
There is already a feature check available via RPC in CRIU. Right now it is only used for pre-copy and post-copy migration as these features are kernel/architecture dependent. For everything else we still rely on version number checks. So far that works for most things.

But adding an interface to libcriu to query the version of the binary would definitely make sense. But we should discuss this upstream.